### ISSUE TITLE
[Ongoing] Sound System. And sound hookups.

### ISSUE THREAD

**Issue by Jaytheway** [TeamPorcupine/ProjectPorcupine#1535]
Since we are not using a third party sound engine, we need to create some functionality for ours. 
I'm not sure what basic Unity Sound features we can already use, so correct me if there is something from this list.

Also here might be a general list with places that should have a hookups for sounds (alternatively we could use a google spreadsheet for this).
### Basic features
- [ ] **Randomize pitch** in range from -X to +Y (semitones, float) every time the sound is played
- [ ] **Randomize volume** in range from -X to _No change_ (dB, float) every time the sound is played (see `dB to Volume` conversion below)
- [ ] **Random container** that randomly choose one sound from the list to be played when event X happens 
- [ ] **Sequence container** that plays next sound from the list when event X happens
- [x] **Distance** volume **attenuation** and **panning** based on the position of a sound emitter

**dB to Volume conversion**
- 6 dB = Double
- -6 dB == 1/2 Volume
- +18 dB == 8x Volume

``` cs
float dBToVolume(float dB)
{
     return powf(10.0f, 0.05f * dB);
}

float VolumeTodB(float volume)
{
    return 20.0f * log10f(volume);
}
```
### Spreadsheet

Here is a spreadsheet that should help to organize our work on the Sound, if anyone wants to contribute in this department: [Sound Design Spreadsheet](https://docs.google.com/spreadsheets/d/1TXwJnrMUbtvmQkYY0zqw5KlBm-eqd-JOlEiWSUjAbOo/edit?usp=sharing)

**Comment by koosemose** [comment]
What is the usecase for random sound and pitch? I can't understand a time when we would want to do that.

**Comment by Jaytheway** [comment]
@koosemose Example: 108 instances of a `SteelWall` being rapidly placed, causing a "machine gun" effect of repeating the same sound many times in a row. Altering pitch and volume of a Sound Effect tricks our brain into thinking that each of this sounds belongs to its own Wall piece. Sounds more natural and less distracting. Same thing with the tiles and any other repeating event.

Other example: if we had a footsteps sound, we wold need to create a 4-7 footstep sound variations (instead of altering the pitch of just a 2-3 of them), because footsteps never sound the same in a real world.

More sound files == more memory usage.
Altering sound parameters (pitch, volume) == less memory usage, very easy for the CPU.

It's a small things, but they add up in a bigger scheme.

**Comment by koosemose** [comment]
Ahh, I see

**Comment by koosemose** [comment]
It might be wise for anyone to hold off implementing anything in the sound system. Problems in the system (#1533) may necessitate it being replaced.

**Comment by koosemose** [comment]
I've got randomization of pitch down, a naive faked container (if you ask for a container you get all audio whose name starts with the passed in name, so `AudioManager.GetAudioSequence("Sound", "Floor_OnCreated")` returns `Floor_OnCreated.ogg`, `Floor_OnCreated1.ogg`, and `Floor_OnCreated2.ogg`) at present you have to manually make use of the container (so choosing a random clip or some way of tracking which number in sequence and using that clip), I also don't know if it's suitable for use in sequence (not sure if they are ordered in a reliable, reasonable, predictable fashion at this time), and simple positional audio (at present if you move the camera while a long sound is playing it won't pan the sound appropriately). This is still a work in progress, so I intend to improve containers and positional audio.

For random volume, I'm not sure if the methods you provided are suited to what I have available to work with. What is the range of dBToVolume? and what is "normal" volume? I need 1.0f to be normal volume, 2.0f to be double volume, and so on, with the ability to go negative (which would invert the signal, which I am not sure if that would be considered desirable)

**Comment by Jaytheway** [comment]
The range of `dBToVolume` is everything above 0. It's basically percentage value for the change of the volume level.
With this formula, I believe, you can not get dBToVolume to be <= 0.
1.0f is current (normal volume), 2.0f is indeed a double. 
Technically it could be 8.0. 
It determines dynamic range of all sounds in the Universe.


**Example**:
 Volume level changes by -12 dB 
 `powf(10.0f, 0.05f * (-12)) = 0.25...`
 Current volume level multiplied by dBToVolume (0.25)

**VolumeTodB example**:
 Volume level lovers by 50% (half of the current value (0.5))
 `20.0f * log10f(0.5) = -6 (dB)`

**Comment by koosemose** [comment]
I'm curious, why is volume randomization from -X to No change? It seems inconsistent behavior from random pitch variation.

**Comment by Geoffrotism** [comment]
@koosemose it could easily be from -X to X, but i think the assumption @Jaytheway was going for was: we set the max volume for the system and then the user can lower the volume from there. Most games start off with Max Volume in all categories and then the user is free to lower them. I personally don't care which way we do it.

**Comment by koosemose** [comment]
@Geoffrotism That's not at all what he was talking about, note I am talking about Volume randomization, his second point, so a sound could be told to have a volume range of, say 6 dB, and it would play anywhere from -6 normal dB to no change in dB, whereas pitch randomization, with a range of 6 semitones would play from -6 to +6 semitones from its normal pitch.

**Comment by Geoffrotism** [comment]
One of these days i need to get some sleep. I didn't even read the header of that line.

**Comment by Jaytheway** [comment]
@koosemose Geoffrotism is partially right. I set the volume levels when rendering audio files with a certain amount of headroom.  Randomization is set after that. I would not be able to consider volume randomization of all sounds (which is varying and should be adjustable) when rendering each individual audio file. 

Basically I set max value when rendering, and to avoid digital clipping, weird unexpected behavior and mix all over the place, we should not make sound clips randomly louder.
